route_generator/src/route_generator.cpp

```diff
@@ -0,0 +1,118 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include <ros/ros.h>
+
+#include "route_generator.h"
+
+RouteGenerator::RouteGenerator(int argc, char **argv, const std::string &name) : route_is_active_(false)
+{
+    ros::init(argc, argv, name);
+}
+
+RouteGenerator::~RouteGenerator() {}
+
+void RouteGenerator::initialize()
+{
+    nh_.reset(new ros::CARMANodeHandle);
+    pnh_.reset(new ros::CARMANodeHandle("~"));
+    nh_->getParam("route_file_path", route_file_path_);
+    route_file_path_pub_ = nh_->advertise<std_msgs::String>("selected_route_path", 1);
+    get_available_route_srv_ = nh_->advertiseService("get_available_routes", &RouteGenerator::get_available_route_cb, this);
+    set_active_route_srv_ = nh_->advertiseService("set_active_route", &RouteGenerator::set_active_route_cb, this);
+    start_active_route_srv_ = nh_->advertiseService("start_active_route", &RouteGenerator::start_active_route_cb, this);
+    abort_active_route_srv_ = nh_->advertiseService("abort_active_route", &RouteGenerator::abort_active_route_cb, this);
+}
+
+void RouteGenerator::run()
+{
+    initialize();   
+    ros::CARMANodeHandle::spin();
+}
+
+bool RouteGenerator::get_available_route_cb(cav_srvs::GetAvailableRoutesRequest &req, cav_srvs::GetAvailableRoutesResponse &resp)
+{
+    ROS_ERROR_STREAM("The call back is triggered!");
+    std::vector<std::string> route_ids = read_route_names(route_file_path_);
+    ROS_ERROR_STREAM("Got a list of names!");
+    ROS_ERROR_STREAM(route_ids.size());
+    for(int i = 0; i < route_ids.size(); ++i)
+    {
+        std::string route_name = route_ids[i].substr(0, route_ids[i].find("."));
+        ROS_ERROR_STREAM(route_name);
+        cav_msgs::Route route_msg;
+        route_msg.routeID = route_name;
+        route_msg.routeName = route_name;
+        route_msg.valid = true;
+        resp.availableRoutes.push_back(route_msg);
+    }
+    ROS_ERROR_STREAM("Callback returned!");
+    return true;
+}
+
+bool RouteGenerator::set_active_route_cb(cav_srvs::SetActiveRouteRequest &req, cav_srvs::SetActiveRouteResponse &resp)
+{
+    if(!route_is_active_)
+    {
+        std::string route_file_name = req.routeID.append(".csv");
+        std_msgs::String selected_route_file_path;
+        selected_route_file_path.data = route_file_path_ + route_file_name;
+        route_file_path_pub_.publish(selected_route_file_path);
+        resp.errorStatus = cav_srvs::SetActiveRouteResponse::NO_ERROR;
+        return true;
+    }
+    return false;
+}
+
+bool RouteGenerator::start_active_route_cb(cav_srvs::StartActiveRouteRequest &req, cav_srvs::StartActiveRouteResponse &resp)
+{
+    if(!route_is_active_)
+    {
+        route_is_active_ = true;
+        resp.errorStatus = cav_srvs::StartActiveRouteResponse::NO_ERROR;
+        return true;
+    }
+    return false;
```

Comment:
Please set the Response status here so we know why the service failed

Reply:
Do this for all services

Reply:
Sounds good. Added the failure case handling.